### Artificial Neural Network for Churning Customer

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('../Building-ANN/Churn_Modelling.csv')

In [3]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Target Variable: <b>Exited </b>

In [4]:
df['Geography'] = pd.get_dummies(df['Geography'],drop_first=True)
df['Gender'] = pd.get_dummies(df['Gender'],drop_first=True)


#### Splitting the data into dependent and Independent variables

In [5]:
X = df.iloc[:,3:13].values
y = df.iloc[:,13].values


#### Splitting the data into Train and Test variables

In [6]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)

#### Scaling the data 

In [7]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

#### Implementing ANN

In [8]:
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


### Implementing the Input and Hidden Layer

In [9]:
classifier = Sequential()
#Adding the input layer and the first hidden layer
classifier.add(Dense(output_dim=6,activation='relu',input_dim=10))

#Adding second hidden layer
classifier.add(Dense(output_dim=6,activation='relu'))

### Adding the output layer:

Updating the activation function using sigmoid function

a. If there are multiple categories, output_dim will be 3,

b. Sigmoid function will be <b>softmax</b>,if the categories are more than 2.


In [10]:
classifier.add(Dense(output_dim=1,init ='uniform',activation='sigmoid'))

### Compiling the ANN

<b>a.Optimizer:</b> The algorithm to find the optimal/best set of weights. Adam is one of the algorithm 

<b>a. Loss function:</b> for NN , 1 neuron - Logistic regression, loss function will be logarithmic loss. 

Binary outcome(0/1)-<b>binary_corossentropy.</b> 
More than 2 categories- it will be <b>categorical_crossentropy</b>.

<b>c. Metrics:</b>  Criterion which we choose to evaluate your metrics.




In [11]:
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

### Fitting the ANN to the training data

batch_size: Number of samples per gradient update.

Number of epoch : An epoch is an iteration over the entire `x` and `y` data provided.

In [12]:
classifier.fit(X_train,y_train,batch_size=10,nb_epoch=100)

Epoch 1/100
8000/8000 [==============================] - 2s 199us/step - loss: 0.5015 - accuracy: 0.7876
Epoch 2/100
8000/8000 [==============================] - 1s 187us/step - loss: 0.4374 - accuracy: 0.7960
Epoch 3/100
8000/8000 [==============================] - 1s 173us/step - loss: 0.4269 - accuracy: 0.8112
Epoch 4/100
8000/8000 [==============================] - 1s 173us/step - loss: 0.4201 - accuracy: 0.8213
Epoch 5/100
8000/8000 [==============================] - 1s 176us/step - loss: 0.4154 - accuracy: 0.8227
Epoch 6/100
8000/8000 [==============================] - 1s 175us/step - loss: 0.4126 - accuracy: 0.8254
Epoch 7/100
8000/8000 [==============================] - 1s 174us/step - loss: 0.4098 - accuracy: 0.8275
Epoch 8/100
8000/8000 [==============================] - 1s 177us/step - loss: 0.4091 - accuracy: 0.8267
Epoch 9/100
8000/8000 [==============================] - 1s 174us/step - loss: 0.4078 - accuracy: 0.8292
Epoch 10/100
8000/8000 [==============================]

8000/8000 [==============================] - 1s 156us/step - loss: 0.3357 - accuracy: 0.8618
Epoch 79/100
8000/8000 [==============================] - 1s 155us/step - loss: 0.3357 - accuracy: 0.8611
Epoch 80/100
8000/8000 [==============================] - 1s 158us/step - loss: 0.3355 - accuracy: 0.8630
Epoch 81/100
8000/8000 [==============================] - 1s 153us/step - loss: 0.3354 - accuracy: 0.8614
Epoch 82/100
8000/8000 [==============================] - 1s 149us/step - loss: 0.3355 - accuracy: 0.8612
Epoch 83/100
8000/8000 [==============================] - 1s 159us/step - loss: 0.3360 - accuracy: 0.8619
Epoch 84/100
8000/8000 [==============================] - 1s 156us/step - loss: 0.3348 - accuracy: 0.8606
Epoch 85/100
8000/8000 [==============================] - 1s 163us/step - loss: 0.3353 - accuracy: 0.8608
Epoch 86/100
8000/8000 [==============================] - 1s 158us/step - loss: 0.3349 - accuracy: 0.8626
Epoch 87/100
8000/8000 [==============================] - 1

### Predicting the Test result

In [13]:
y_pred = classifier.predict(X_test)

In [14]:
y_pred = (y_pred>0.5)


In [15]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)
print('\n Confusion Matrix \n',cm)


 Confusion Matrix 
 [[1501   94]
 [ 185  220]]


### Miscellaneous:
    
New Customer with single prediction if a new customer will leave the bank or not. Data is as below:

Geography: France
    
Credit Score: 600
    
Gender: Male
    
Age: 40 years old
    
Tenure: 3 years
    
Balance: $60000
    
Number of Products: 2
    
Does this customer have a credit card ? Yes

Is this customer an Active Member: Yes
    
Estimated Salary: $50000
    
    

In [16]:
new_pred = classifier.predict(sc.transform(np.array([[600,0,0,40,3,60000,2,1,1,50000]])))
new_pred = (new_pred>0.5)

In [28]:
print('Prediction for new customer data will be',new_pred[0])

Prediction for new customer data will be [False]


### Conclusion:
    
    
    This marks the end of our basic Artificial Neural Network with 86% accuracy for predicting the customer churn. 
    In the next excersise we will use this model to tune the accuracy to improve the prediction.